src/CoreBundle/EventListener/CourseAccessListener.php

```diff
@@ -7,41 +7,80 @@
 namespace Chamilo\CoreBundle\EventListener;
 
 use Chamilo\CoreBundle\Entity\TrackECourseAccess;
-use Chamilo\CourseBundle\Event\CourseAccess;
-use Doctrine\ORM\EntityManager;
-use Symfony\Component\HttpFoundation\Request;
+use Chamilo\CoreBundle\Entity\User;
+use Chamilo\CoreBundle\ServiceHelper\CidReqHelper;
+use DateTime;
+use Symfony\Component\HttpKernel\Event\RequestEvent;
+use Doctrine\ORM\EntityManagerInterface;
 use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\Security\Core\Authentication\Token\Storage\TokenStorageInterface;
 
 /**
  * In and outs of a course
  * This listener is always called when user enters the course home.
  */
 class CourseAccessListener
 {
-    protected ?Request $request = null;
-
     public function __construct(
-        private readonly EntityManager $em,
-        RequestStack $requestStack
-    ) {
-        $this->request = $requestStack->getCurrentRequest();
+        private readonly EntityManagerInterface $em,
+        private readonly RequestStack $requestStack,
+        private readonly CidReqHelper $cidReqHelper,
+        private readonly TokenStorageInterface $tokenStorage,
+    ) {}
+
+    public function onKernelRequest(RequestEvent $event): void
+    {
+        if (!$event->isMainRequest() || $event->getRequest()->attributes->get('access_checked')) {
+            // If it's not the main request or we've already handled access in this request, do nothing
+            return;
+        }
+
+        $courseId = (int) $this->cidReqHelper->getCourseId();
+        $sessionId = (int) $this->cidReqHelper->getSessionId();
+
+        if ($courseId > 0) {
+            $user = $this->tokenStorage->getToken()?->getUser();
+            if ($user instanceof User) {
+                $ip = $this->requestStack->getCurrentRequest()->getClientIp();
```

Comment:
La request puedes obtenerla desde el mismo envento

```suggestion
                $ip = $event->getRequest()->getClientIp();
```